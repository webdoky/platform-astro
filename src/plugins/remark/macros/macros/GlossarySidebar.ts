import type { Root } from 'mdast';
import { SKIP } from 'unist-util-visit';

import getChildren from '../../../registry/get-children.ts';
import type { RawPage } from '../../../registry/validation.ts';
import type { AstroFile } from '../../validate-astro-file.ts';
import type { MacroFunction } from '../types.ts';

function macro(_tree: Root, file: AstroFile) {
  if (file.data.astro.frontmatter.sidebar) {
    throw new Error('Sidebar already exists');
  }
  const targetLocale = process.env.TARGET_LOCALE;
  const glossaryPages = getChildren('Glossary');
  function sortByTitle({ title: titleA = '' }, { title: titleB = '' }) {
    return titleA.localeCompare(titleB, targetLocale);
  }

  const pageToNavItem = ({ slug, title }: RawPage, currentSlug: string) => ({
    title,
    path: `/${targetLocale}/docs/${slug}/`,
    hasTranslation: true,
    isCurrent: slug === currentSlug,
  });
  // TODO: we definitely need it typed at some point
  file.data.astro.frontmatter.sidebar = [
    {
      links: [],
      sections: [
        {
          expanded: true,
          items: glossaryPages
            .map((item) =>
              pageToNavItem(item, file.data.astro.frontmatter.slug),
            )
            .sort(sortByTitle),
          title: 'Терміни глосарія',
        },
      ],
    },
  ];
}

const GlossarySidebar: MacroFunction = (_node, index, parent, tree, file) => {
  macro(tree, file);
  parent.children.splice(index, 1);
  return [SKIP, index];
};

export default GlossarySidebar;
