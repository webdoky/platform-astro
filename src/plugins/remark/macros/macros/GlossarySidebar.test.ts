import type { Root } from 'mdast';
import { afterAll, beforeEach, describe, expect, it, vi } from 'vitest';

import GlossarySidebar from './GlossarySidebar.ts';

vi.mock('../../../registry/get-children.ts', () => ({
  default: vi.fn().mockReturnValue([
    {
      title: 'Хибні значення',
      slug: 'Glossary/Falsy',
    },
    {
      hasLocalizedContent: true,
      title: 'Істинні значення',
      slug: 'Glossary/Truthy',
    },
    {
      title: 'Undefined',
      slug: 'Glossary/Undefined',
    },
    {
      title: 'Null',
      slug: 'Glossary/Null',
    },
    {
      title: 'NaN',
      slug: 'Glossary/NaN',
    },
    {
      title: 'Primitive',
      slug: 'Glossary/Primitive',
    },
    {
      title: 'Object',
      slug: 'Glossary/Object',
    },
    {
      title: 'Array',
      slug: 'Glossary/Array',
    },
    {
      title: 'Function',
      slug: 'Glossary/Function',
    },
    {
      title: 'Operator',
      slug: 'Glossary/Operator',
    },
    {
      title: 'Statement',
      slug: 'Glossary/Statement',
    },
    {
      title: 'Expression',
      slug: 'Glossary/Expression',
    },
    {
      title: 'Keyword',
      slug: 'Glossary/Keyword',
    },
    {
      title: 'Property',
      slug: 'Glossary/Property',
    },
    {
      title: 'Event',
      slug: 'Glossary/Event',
    },
    {
      title: 'Method',
      slug: 'Glossary/Method',
    },
    {
      title: 'Interface',
      slug: 'Glossary/Interface',
    },
  ]),
}));

describe('GlossarySidebar', () => {
  const OLD_ENV = process.env;

  beforeEach(() => {
    vi.resetModules(); // Most important - it clears the cache
    process.env = {
      TARGET_LOCALE: 'uk',
    };
  });

  afterAll(() => {
    process.env = OLD_ENV; // Restore old environment
  });
  it('should throw an error if the sidebar already exists', () => {
    expect(() => {
      const tree = {
        type: 'root',
        children: [
          {
            type: 'macro',
            name: 'GlossarySidebar',
            parameters: [],
          },
        ],
      };
      const file = {
        data: {
          astro: {
            frontmatter: {
              sidebar: [],
              'page-type': '',
              title: '',
              slug: 'Glossary/Undefined',
            },
          },
        },
      } as any;
      GlossarySidebar(tree as Root, file);
    }).toThrowError('Sidebar already exists');
  });
  it('should add a sidebar to the frontmatter', () => {
    const tree = {
      type: 'root',
      children: [
        {
          type: 'macro',
          name: 'GlossarySidebar',
          parameters: [],
        },
      ],
    };
    const file = {
      data: {
        astro: {
          frontmatter: {},
        },
      },
    } as any;
    GlossarySidebar(tree as Root, file);
    expect(file.data.astro.frontmatter.sidebar).toMatchSnapshot();
  });
});
