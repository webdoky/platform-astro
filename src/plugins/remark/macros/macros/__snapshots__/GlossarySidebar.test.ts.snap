// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GlossarySidebar > should add a sidebar to the frontmatter 1`] = `
[
  {
    "links": [],
    "sections": [
      {
        "expanded": true,
        "items": [
          {
            "hasTranslation": true,
            "isCurrent": false,
            "path": "/uk/docs/Glossary/Truthy",
            "title": "Істинні значення",
          },
          {
            "hasTranslation": true,
            "isCurrent": false,
            "path": "/uk/docs/Glossary/Falsy",
            "title": "Хибні значення",
          },
          {
            "hasTranslation": true,
            "isCurrent": false,
            "path": "/uk/docs/Glossary/Array",
            "title": "Array",
          },
          {
            "hasTranslation": true,
            "isCurrent": false,
            "path": "/uk/docs/Glossary/Event",
            "title": "Event",
          },
          {
            "hasTranslation": true,
            "isCurrent": false,
            "path": "/uk/docs/Glossary/Expression",
            "title": "Expression",
          },
          {
            "hasTranslation": true,
            "isCurrent": false,
            "path": "/uk/docs/Glossary/Function",
            "title": "Function",
          },
          {
            "hasTranslation": true,
            "isCurrent": false,
            "path": "/uk/docs/Glossary/Interface",
            "title": "Interface",
          },
          {
            "hasTranslation": true,
            "isCurrent": false,
            "path": "/uk/docs/Glossary/Keyword",
            "title": "Keyword",
          },
          {
            "hasTranslation": true,
            "isCurrent": false,
            "path": "/uk/docs/Glossary/Method",
            "title": "Method",
          },
          {
            "hasTranslation": true,
            "isCurrent": false,
            "path": "/uk/docs/Glossary/NaN",
            "title": "NaN",
          },
          {
            "hasTranslation": true,
            "isCurrent": false,
            "path": "/uk/docs/Glossary/Null",
            "title": "Null",
          },
          {
            "hasTranslation": true,
            "isCurrent": false,
            "path": "/uk/docs/Glossary/Object",
            "title": "Object",
          },
          {
            "hasTranslation": true,
            "isCurrent": false,
            "path": "/uk/docs/Glossary/Operator",
            "title": "Operator",
          },
          {
            "hasTranslation": true,
            "isCurrent": false,
            "path": "/uk/docs/Glossary/Primitive",
            "title": "Primitive",
          },
          {
            "hasTranslation": true,
            "isCurrent": false,
            "path": "/uk/docs/Glossary/Property",
            "title": "Property",
          },
          {
            "hasTranslation": true,
            "isCurrent": false,
            "path": "/uk/docs/Glossary/Statement",
            "title": "Statement",
          },
          {
            "hasTranslation": true,
            "isCurrent": false,
            "path": "/uk/docs/Glossary/Undefined",
            "title": "Undefined",
          },
        ],
        "title": "Терміни глосарія",
      },
    ],
  },
]
`;
